File: app/src/main/java/se/accidis/fmfg/app/services/DocumentsRepository.java
package se.accidis.fmfg.app.services;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import org.joda.time.DateTime;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

import se.accidis.fmfg.app.model.Document;
import se.accidis.fmfg.app.model.DocumentLink;
import se.accidis.fmfg.app.model.DocumentSettings;
import se.accidis.fmfg.app.utils.IOUtils;

/**
 * Repository for documents.
 */
public final class DocumentsRepository {
    private static final String CURRENT_DOCUMENT = "CurrentDocument.json";
    private static final String SAVED_DOCUMENT_FORMAT = "Saved_%1$s.json";
    private static final String SAVED_DOCUMENT_PREFIX = "Saved_";
    private static final String TAG = DocumentsRepository.class.getSimpleName();
    private final Context mContext;
    private Document mCurrentDocument;
    private static DocumentsRepository mInstance;
    private List<DocumentLink> mList;
    private OnLoadedListener mOnLoadedListener;

    private DocumentsRepository(Context context) {
        mContext = context.getApplicationContext();
    }

    public static DocumentsRepository getInstance(Context context) {
        return (null == mInstance ? (mInstance = new DocumentsRepository(context)) : mInstance);
    }

    public void beginLoad() {
        if (isLoaded()) {
            Log.d(TAG, "Documents already loaded, nothing to do.");
            if (null != mOnLoadedListener) {
                mOnLoadedListener.onLoaded(mList);
            }

            return;
        }

        Log.d(TAG, "Loading documents.");
        LoadTask loadTask = new LoadTask();
        loadTask.execute();
    }

    public void changeCurrentDocument(Document document) {
        Log.d(TAG, "Replacing current document with ID: " + document.getId());
        mCurrentDocument = document;
        commitCurrentDocument();
    }

    public void commitCurrentDocument() {
        try {
            String json = mCurrentDocument.toJson().toString();
            IOUtils.writeToStream(mContext.openFileOutput(CURRENT_DOCUMENT, Context.MODE_PRIVATE), json);
        } catch (Exception ex) {
            Log.e(TAG, "Exception while writing current document.", ex);
        }
    }

    public void deleteDocument(UUID id) {
        Log.d(TAG, "Deleting document with ID: " + id);
        String filename = getFilenameByDocumentId(id);
        mContext.deleteFile(filename);
        invalidate();
    }

    public void ensureCurrentDocumentLoaded() {
        if (null == mCurrentDocument) {
            try {
                mCurrentDocument = readDocument(CURRENT_DOCUMENT);
            } catch (FileNotFoundException ignored) {
                Log.w(TAG, "Current document not found, might be first startup.");
            } catch (Exception ex) {
                Log.e(TAG, "Exception while reading current document.", ex);
            }

            if (null == mCurrentDocument) {
                mCurrentDocument = new Document();
                Log.d(TAG, "Created a document with ID: " + mCurrentDocument.getId());
            } else {
                Log.d(TAG, "Loaded current document with ID: " + mCurrentDocument.getId());
            }
        }
    }

    public Document getCurrentDocument() {
        ensureCurrentDocumentLoaded();
        return mCurrentDocument;
    }

    public boolean isLoaded() {
        return (null != mList);
    }

    public Document loadDocument(UUID id) throws IOException, JSONException {
        Log.d(TAG, "Loading document with ID: " + id);
        String filename = getFilenameByDocumentId(id);
        return readDocument(filename);
    }

    public void saveCurrentDocument(String name) throws IOException, JSONException {
        Log.d(TAG, "Saving current document with ID: " + mCurrentDocument.getId() + ", name = " + mCurrentDocument.getName());
        ensureCurrentDocumentLoaded();
        mCurrentDocument.setName(name);
        mCurrentDocument.setTimestamp(DateTime.now());
        mCurrentDocument.getSettings().remove(DocumentSettings.Keys.UNSAVED_CHANGES);
        writeDocument(mCurrentDocument);
        invalidate();
    }

    public void setOnLoadedListener(OnLoadedListener listener) {
        mOnLoadedListener = listener;
    }

    private static String getFilenameByDocumentId(UUID documentId) {
        return String.format(SAVED_DOCUMENT_FORMAT, documentId.toString());
    }

    private void invalidate() {
        mList = null;
    }

    private Document readDocument(String fileName) throws IOException, JSONException {
        String str = IOUtils.readToEnd(mContext.openFileInput(fileName));
        JSONObject json = new JSONObject(str);
        return Document.fromJson(json);
    }

    private void writeDocument(Document document) throws IOException, JSONException {
        String fileName = getFilenameByDocumentId(document.getId());
        String json = document.toJson().toString();
        IOUtils.writeToStream(mContext.openFileOutput(fileName, Context.MODE_PRIVATE), json);
    }

    public interface OnLoadedListener {
        void onException(Exception ex);

        void onLoaded(List<DocumentLink> list);
    }

    private final class LoadTask extends AsyncTask<Void, Void, Void> {
        @Override
        protected Void doInBackground(Void... params) {
            try {
                List<String> fileList = filterDocuments(mContext.fileList());

                ArrayList<DocumentLink> list = new ArrayList<>();
                for (String file : fileList) {
                    DocumentLink docLink = readDocumentLink(file);
                    list.add(docLink);
                }

                Collections.sort(list);
                mList = list;

                Log.i(TAG, "Finished loading documents (" + mList.size() + " documents loaded).");

            } catch (Exception ex) {
                Log.e(TAG, "Failed to load documents.", ex);

                if (null == mList && null != mOnLoadedListener) {
                    mOnLoadedListener.onException(ex);
                }
            }

            if (null != mList && null != mOnLoadedListener) {
                mOnLoadedListener.onLoaded(mList);
            }

            return null;
        }

        private List<String> filterDocuments(String[] files) {
            List<String> filtered = new ArrayList<>();
            for (String fileName : files) {
                if (fileName.startsWith(SAVED_DOCUMENT_PREFIX)) {
                    filtered.add(fileName);
                }
            }
            return filtered;
        }

        private DocumentLink readDocumentLink(String fileName) throws IOException, JSONException {
            String str = IOUtils.readToEnd(mContext.openFileInput(fileName));
            JSONObject json = new JSONObject(str);
            return DocumentLink.fromJson(json);
        }
    }
}

File: app/src/main/java/se/accidis/fmfg/app/services/DocumentsRepository.kt
package se.accidis.fmfg.app.services

import android.content.Context
import android.content.Context.MODE_PRIVATE
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject
import se.accidis.fmfg.app.model.*
import se.accidis.fmfg.app.utils.Resource
import timber.log.Timber
import java.io.FileNotFoundException
import java.util.UUID

/**
 * Repository for documents.
 */
class DocumentsRepository private constructor(private val context: Context) {
    private val prefs = Preferences(context)

    private val _currentDocument = MutableStateFlow<Document?>(null)
    val currentDocumentFlow: StateFlow<Document>
        get() {
            ensureDocument()
            @Suppress("UNCHECKED_CAST")
            return _currentDocument as StateFlow<Document>
        }

    val currentDocument: Document
        get() = ensureDocument()

    private val _documents = MutableStateFlow<Resource<List<DocumentLink>>>(Resource.Loading)
    val documents: StateFlow<Resource<List<DocumentLink>>> = _documents.asStateFlow()

    private val repositoryScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    fun beginLoad() {
        val current = _documents.value
        if (current is Resource.Success && current.data.isNotEmpty()) {
            Timber.d("Documents already loaded, nothing to do.")
            return
        }

        Timber.d("Loading documents.")
        _documents.value = Resource.Loading
        repositoryScope.launch {
            try {
                val list = withContext(Dispatchers.IO) {
                    context.fileList()
                        .filter { it.startsWith(SAVED_DOCUMENT_PREFIX) }
                        .map { readDocumentLink(it) }
                        .sorted()
                }

                _documents.value = Resource.Success(list)
                Timber.i("Finished loading documents (%d documents loaded).", list.size)
            } catch (ex: Exception) {
                Timber.e(ex, "Failed to load documents.")
                _documents.value = Resource.Error(ex)
            }
        }
    }

    val isLoaded: Boolean
        get() = _documents.value is Resource.Success

    fun updateCurrentDocument(document: Document) {
        Timber.d("Updating current document with ID: %s", document.id)
        _currentDocument.value = document
        repositoryScope.launch(Dispatchers.IO) {
            try {
                val json = document.toJson().toString()
                context.openFileOutput(CURRENT_DOCUMENT, MODE_PRIVATE).bufferedWriter().use {
                    it.write(json)
                }
            } catch (ex: Exception) {
                Timber.e(ex, "Exception while writing current document.")
            }
        }
    }

    fun saveCurrentDocument(name: String) {
        val document = ensureDocument()
        Timber.d("Saving current document with ID: %s, name = %s", document.id, name)
        val savedDoc = document.mutate { this.name = name }.save()
        repositoryScope.launch(Dispatchers.IO) {
            writeDocument(savedDoc)
            invalidateListOfDocuments()
        }
    }

    suspend fun loadDocument(id: UUID): Document = withContext(Dispatchers.IO) {
        Timber.d("Loading document with ID: %s", id)
        val filename = getFilenameByDocumentId(id)
        readDocument(filename)
    }

    fun deleteDocument(id: UUID) {
        Timber.d("Deleting document with ID: %s", id)
        repositoryScope.launch(Dispatchers.IO) {
            val filename = getFilenameByDocumentId(id)
            context.deleteFile(filename)
            invalidateListOfDocuments()
        }
    }

    private fun ensureDocument(): Document {
        val current = _currentDocument.value
        if (current != null) return current

        val document = try {
            // Using runBlocking here for synchronous initialization when called from property getter
            runBlocking(Dispatchers.IO) {
                readDocument(CURRENT_DOCUMENT)
            }
        } catch (_: FileNotFoundException) {
            Timber.w("Current document not found, might be first startup.")
            null
        } catch (ex: Exception) {
            Timber.e(ex, "Exception while reading current document.")
            null
        } ?: DocumentBuilder.createNew().mutate { author = prefs.defaultAuthor }

        _currentDocument.value = document
        Timber.d("Current document initialized with ID: %s", document.id)
        return document
    }

    private fun invalidateListOfDocuments() {
        _documents.value = Resource.Loading
        beginLoad()
    }

    private fun readDocument(fileName: String): Document {
        val str = context.openFileInput(fileName).bufferedReader().use { it.readText() }
        return Document.fromJson(JSONObject(str))
    }

    private fun writeDocument(document: Document) {
        val fileName = getFilenameByDocumentId(document.id)
        val json = document.toJson().toString()
        context.openFileOutput(fileName, Context.MODE_PRIVATE).bufferedWriter().use {
            it.write(json)
        }
    }

    private fun readDocumentLink(fileName: String): DocumentLink {
        val str = context.openFileInput(fileName).bufferedReader().use { it.readText() }
        return DocumentLink.fromJson(JSONObject(str))
    }

    companion object {
        private const val CURRENT_DOCUMENT = "CurrentDocument.json"
        private const val SAVED_DOCUMENT_PREFIX = "Saved_"

        @Volatile
        private var singleton: DocumentsRepository? = null

        @JvmStatic
        fun getInstance(context: Context): DocumentsRepository =
            singleton ?: synchronized(this) {
                singleton ?: DocumentsRepository(context).also { singleton = it }
            }

        private fun getFilenameByDocumentId(documentId: UUID): String =
            "Saved_$documentId.json"
    }
}
